File: src/main/java/com/meti/AbstractStream.java
package com.meti;

public abstract class AbstractStream<T> implements Stream<T> {
	@Override
	public <R> Stream<R> mapE1(F1E1<T, R, ?> mapper) {
		return new AbstractStream<R>() {
			@Override
			public R head() throws StreamException {
				try {
					return mapper.apply(AbstractStream.this.head());
				} catch (EndOfStreamException e) {
					throw e;
				} catch (Exception e) {
					throw new StreamException(e);
				}
			}
		};
	}

	@Override
	public boolean allMatch(F1<T, Boolean> predicate) throws StreamException {
		while (true) {
			try {
				var apply = head();
				if (!predicate.apply(apply)) {
					return false;
				}
			} catch (EndOfStreamException e) {
				break;
			}
		}
		return true;
	}
}

File: src/main/java/com/meti/AbstractStream.scala
package com.meti

import java.lang

abstract class AbstractStream[T] extends Stream[T] {
  override def fold[R](initial: R, folder: F2[R, T, R]): R = ???

  override def fold(folder: F2[T, T, T]): Option[T] = ???

  override def map[R](mapper: F1[T, R]): Stream[R] = ???

  override def allMatch(predicate: F1[T, lang.Boolean]): Boolean = {
    var continue = true
    while (continue)
      try if (!predicate.apply(head))
        return false
      catch {
        case _: EndOfStreamException => continue = false
      }
    true
  }

  override def mapE1[R](mapper: F1E1[T, R, _]): AbstractStream[R] = new AbstractStream[R]() {
    @throws[StreamException]
    override def head: R = try mapper.apply(AbstractStream.this.head())
    catch {
      case e: EndOfStreamException => throw e
      case e: Exception => throw new StreamException(e)
    }
  }
}